java/src/jmri/jmrix/AbstractThrottle.java

```diff
@@ -836,378 +836,452 @@ public float getSpeedIncrement() {
      */
     @Override
     public void setF21(boolean f21) {
-        boolean old = this.f21;
-        this.f21 = f21;
+        updateFunction(21,f21);
         sendFunctionGroup5();
-        if (old != this.f21) {
-            notifyPropertyChangeListener(Throttle.F21, old, this.f21);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF22(boolean f22) {
-        boolean old = this.f22;
-        this.f22 = f22;
+        updateFunction(22,f22);
         sendFunctionGroup5();
-        if (old != this.f22) {
-            notifyPropertyChangeListener(Throttle.F22, old, this.f22);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF23(boolean f23) {
-        boolean old = this.f23;
-        this.f23 = f23;
+        updateFunction(23,f23);
         sendFunctionGroup5();
-        if (old != this.f23) {
-            notifyPropertyChangeListener(Throttle.F23, old, this.f23);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF24(boolean f24) {
-        boolean old = this.f24;
-        this.f24 = f24;
+        updateFunction(24,f24);
         sendFunctionGroup5();
-        if (old != this.f24) {
-            notifyPropertyChangeListener(Throttle.F24, old, this.f24);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF25(boolean f25) {
-        boolean old = this.f25;
-        this.f25 = f25;
+        updateFunction(25,f25);
         sendFunctionGroup5();
-        if (old != this.f25) {
-            notifyPropertyChangeListener(Throttle.F25, old, this.f25);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF26(boolean f26) {
-        boolean old = this.f26;
-        this.f26 = f26;
+        updateFunction(26,f26);
         sendFunctionGroup5();
-        if (old != this.f26) {
-            notifyPropertyChangeListener(Throttle.F26, old, this.f26);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF27(boolean f27) {
-        boolean old = this.f27;
-        this.f27 = f27;
+        updateFunction(27,f27);
         sendFunctionGroup5();
-        if (old != this.f27) {
-            notifyPropertyChangeListener(Throttle.F27, old, this.f27);
-        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void setF28(boolean f28) {
-        boolean old = this.f28;
-        this.f28 = f28;
+        updateFunction(28,f28);
         sendFunctionGroup5();
-        if (old != this.f28) {
-            notifyPropertyChangeListener(Throttle.F28, old, this.f28);
+    }
+    
+    /**
+     * Update the state of a single function.
+     * Updates function value and ChangeListener.
+     * Does not send outward message TO hardware.
+     * @param fn Function Number 0-28
+     * @param state On - True, Off - False
+     */
+    public void updateFunction(int fn, boolean state) {
```

Comment:
I have mixed emotions about this change.

You've taken a set of small functions and combined their logic into one.  This is the opposite of the way good coding practice should have us moving.  Codeclimate correctly flags this by noting
- The updateFunction method is Long (179 lines) which makes it difficult to understand
- The updateFunction method has a large cognitive complexity (59) which exceeds  

But, I see that you're taking this from several places and putting it into one, and that generally is a good thing.

But I think we can do better, and here's how:
Change the f0..f28 variables in AbstractThrottle to be an array with 29 entries

re-write this method so it uses the array and constructs the property change values instead of using the constants.  (see the jmri.Throttle interface which defines them)

You'll also need to modify any throttle that extends AbstractThrottle and uses the f0..f28 values directly, which increase the footprint of this PR, but the code will be cleaner.  (These are protected values, so they aren't externally visible, except in the AbstractThrottle and it's decendents ).

Reply:
Incidentally, if you want to get this merged in as is, I would be happy to take on the changes I suggest.

Reply:
Thanks for the comments, I was contemplating an Enum for the functions but an Array much easier, please see update.